apollo-router-core/src/spec/query.rs

```diff
@@ -176,79 +194,68 @@ impl Query {
-                        self.apply_selection_set(selection_set, input, output);
+                    if let Some(selection_set) = self
+                        .fragments
+                        .get(name)
+                        .or_else(|| schema.fragments.get(name))
+                    {
+                        self.apply_selection_set(selection_set, input, output, schema);
                     } else {
                         failfast_debug!("Missing fragment named: {}", name);
                     }
                 }
             }
         }
     }
-}
 
-#[derive(Debug, Clone)]
-pub(crate) enum Selection {
-    Field {
-        name: String,
-        selection_set: Option<Vec<Selection>>,
-    },
-    InlineFragment {
-        selection_set: Vec<Selection>,
-    },
-    FragmentSpread {
-        name: String,
-    },
-}
-
-impl From<ast::Selection> for Selection {
-    fn from(selection: ast::Selection) -> Self {
-        match selection {
-            // Spec: https://spec.graphql.org/draft/#Field
-            ast::Selection::Field(field) => {
-                let name = field
-                    .name()
-                    .expect("the node Name is not optional in the spec; qed")
-                    .text()
-                    .to_string();
-                let alias = field.alias().map(|x| x.name().unwrap().text().to_string());
-                let name = alias.unwrap_or(name);
-                let selection_set = field
-                    .selection_set()
-                    .map(|x| x.selections().into_iter().map(Into::into).collect());
+    /// Validate a [`Request`]'s variables against this [`Query`] using a provided [`Schema`].
+    #[tracing::instrument(skip_all, level = "trace")]
+    pub fn validate_variables(&self, request: &Request, schema: &Schema) -> Result<(), Response> {
+        let operation_name = request.operation_name.as_deref();
+        let operation_variable_types =
+            self.operations
+                .iter()
+                .fold(HashMap::new(), |mut acc, operation| {
+                    if operation_name.is_none() || operation.name.as_deref() == operation_name {
+                        acc.extend(operation.variables.iter())
+                    }
+                    acc
+                });
 
-                Self::Field {
-                    name,
-                    selection_set,
-                }
-            }
-            // Spec: https://spec.graphql.org/draft/#InlineFragment
-            ast::Selection::InlineFragment(inline_fragment) => {
-                let selection_set = inline_fragment
-                    .selection_set()
-                    .expect("the node SelectionSet is not optional in the spec; qed")
-                    .selections()
-                    .into_iter()
-                    .map(Into::into)
-                    .collect();
-
-                Self::InlineFragment { selection_set }
+        if LevelFilter::current() >= LevelFilter::DEBUG {
```

Comment:
could this entire block be gated by the `failfast` feature?

Reply:
Right now it's a mix of the 2:
- without failfast it prints a DEBUG warning only if the debug level is active
- with failfast and if the debug level is active, it panicks

So the performance hit is only if using level >= DEBUG

I think it's best this way but I don't mind changing it to failfast only.

Reply:
I do not think a panic is a behaviour we want, even in debug. If someone wants to see more logs in a production system and activates debug logs, the expectation is that it should not trigger more panics

Reply:
The panic is only when you activate the feature failfast. No panic if you use log level debug without failfast

Reply:
https://github.com/apollographql/router/blob/13045c514ec9954f2c3713999c31b4b75fa805e9/apollo-router-core/src/lib.rs#L3-L12

Reply:
alright, I see now, thanks